ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiPersistedWorldState.java

```diff
@@ -104,43 +104,76 @@ public BonsaiWorldStateKeyValueStorage getWorldStateStorage() {
   protected Hash calculateRootHash(
       final BonsaiWorldStateKeyValueStorage.BonsaiUpdater stateUpdater,
       final BonsaiWorldStateUpdater worldStateUpdater) {
-    // first clear storage
-    for (final Address address : worldStateUpdater.getStorageToClear()) {
-      // because we are clearing persisted values we need the account root as persisted
-      final BonsaiAccount oldAccount =
-          worldStateStorage
-              .getAccount(Hash.hash(address))
-              .map(bytes -> fromRLP(BonsaiPersistedWorldState.this, address, bytes, true))
-              .orElse(null);
-      if (oldAccount == null) {
-        // This is when an account is both created and deleted within the scope of the same
-        // block.  A not-uncommon DeFi bot pattern.
-        continue;
+    clearStorage(stateUpdater, worldStateUpdater);
+
+    // This must be done before updating the accounts so
+    // that we can get the storage state hash
+    updateAccountStorageState(stateUpdater, worldStateUpdater);
```

Comment:
In a next PR it will be nice to add also this modification here https://github.com/hyperledger/besu/pull/4568/files#diff-0cb922a2c0562efbff99afb7bbde477d58a8adb52a82b076756ac0d90e387fe6R58

Reply:
I'm going to make a next PR to apply the same modifications to BonsaiPersistedWorldState.